velox/exec/Window.cpp

```diff
@@ -379,25 +436,85 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        if (frameArg->index == kConstantChannel) {
+          auto constantOffset =
+              frameArg->value->template as<ConstantVector<int64_t>>()->valueAt(
+                  0);
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow - constantOffset - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          auto offsetsVector = frameArg->value->asFlatVector<int64_t>();
+          for (auto i = 0; i < numRows; i++) {
+            rawFrameBounds[i] =
+                (startRow + i) - offsetsVector->valueAt(i) - firstPartitionRow;
+          }
+        }
+        break;
+      }
+      case core::WindowNode::BoundType::kFollowing: {
+        if (frameArg->index == kConstantChannel) {
+          auto constantOffset =
+              frameArg->value->template as<ConstantVector<int64_t>>()->valueAt(
+                  0);
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow + constantOffset - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          auto offsetsVector = frameArg->value->asFlatVector<int64_t>();
+          for (auto i = 0; i < numRows; i++) {
+            rawFrameBounds[i] =
+                (startRow + i) + offsetsVector->valueAt(i) - firstPartitionRow;
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  auto fixOutOfBoundFrames = [&](vector_size_t* frameStart,
+                                 vector_size_t* frameEnd,
+                                 vector_size_t lastRow) -> void {
+    for (auto i = frameStart, j = frameEnd; i < frameStart + numRows;
+         i++, j++) {
+      if (*i <= *j) {
+        *i = std::max(*i, 0);
+        *j = std::min(*j, lastRow);
+      } else {
+        *i = -1;
+        *j = -1;
+      }
+    }
+  };
+
   for (auto i = 0; i < numFuncs; i++) {
     updateFrameBounds(
         rawFrameStartBuffers[i],
         windowFrames_[i].startType,
         windowFrames_[i].type,
-        true);
+        true,
+        windowFrames_[i].startFrameArg);
     updateFrameBounds(
         rawFrameEndBuffers[i],
         windowFrames_[i].endType,
         windowFrames_[i].type,
-        false);
+        false,
+        windowFrames_[i].endFrameArg);
+  }
```

Comment:
It might be simpler to add the fixOutOfBoundFrames in this loop itself. Seems like that check is applicable only if either bound type has kPreceding or kFollowing. So if either had a FrameChannelArg then the out of bound check is applicable.

The hasOutOfBoundFrames_ boolean is not needed.

Reply:
Thanks for the suggestion, it was in a separate loop to avoid checking in the loop for other frames. Changed accordingly!